### PR TITLE
feat(openwith): Add support for the box-edit integration

### PR DESCRIPTION
**Pull request by jeremypress**

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @jeremypress has signed the CLA. Thanks for the pull request!

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 1]
```diff
@@ -194,7 +203,57 @@ class ContentOpenWith extends PureComponent<Props, State> {
...
+        });
+
+        if (boxEditIntegration && !boxEditIntegration.isDisabled) {
+            this.boxEdit = new BoxEdit();
+            // Check if Box Tools is installed and reachable
+            this.boxEdit
+                .checkBoxEditAvailability()
+                .catch(() => {
+                    boxEditIntegration.isDisabled = true;
+                    boxEditIntegration.disabledReasons = ['Install Box Tools to open this file on your desktop'];
```
Comment (jeremypress) [inline src/components/ContentOpenWith/ContentOpenWith.js]: Will become a localized string

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 2]
```diff
@@ -194,7 +203,57 @@ class ContentOpenWith extends PureComponent<Props, State> {
...
+                        return;
+                    }
+
+                    this.boxEdit
+                        // Ping box edit server to see if we can open this file
+                        .canOpenWithBoxEdit([extension])
+                        .catch(() => {
+                            // Box Tools was unable to open this file type
+                            boxEditIntegration.isDisabled = true;
+                            boxEditIntegration.disabledReasons = ['Box Tools cannot open this file type'];
```
Comment (jeremypress) [inline src/components/ContentOpenWith/ContentOpenWith.js]: Will also become a localized string

File: src/components/ContentOpenWith/OpenWithButton.js [hunk 3]
```diff
@@ -29,32 +29,53 @@ type Props = {
...
+    disabledReasons: Array<string> = [],
+    displayDescription: string,
+): ?(string | Element) => {
     if (isLoading) {
         return null;
     }
+
     if (error) {
-        return <FormattedMessage {...messages.errorOpenWithDescription} />;
+        return disabledReasons[0] || <FormattedMessage {...messages.errorOpenWithDescription} />;
```
Comment (jstoffan) [inline src/components/ContentOpenWith/OpenWithButton.js]: Lots of duplication here. Would this work?

```
if (isLoading) {
    return null;
}

if (disabledReasons[0]) {
    return disabledReasons[0];
}

//etc.
```

File: src/components/ContentOpenWith/OpenWithButton.js [hunk 4]
```diff
@@ -29,32 +29,53 @@ type Props = {
...
 const OpenWithButton = ({ error, onClick, displayIntegration, isLoading }: Props) => {
-    const { displayDescription, displayName, isDisabled: isDisplayIntegrationDisabled } = displayIntegration || {};
+    const { displayName, isDisabled: isDisplayIntegrationDisabled, extension, disabledReasons, displayDescription } =
+        displayIntegration || {};
 
     const isDisabled = !!isDisplayIntegrationDisabled || !displayName;
-    const IntegrationIcon = displayName && ICON_FILE_MAP[displayName];
-    const Icon = IntegrationIcon || IconOpenWith;
+
+    let Icon;
```
Comment (jstoffan) [inline src/components/ContentOpenWith/OpenWithButton.js]: What's the reason for this change?

File: src/components/ContentOpenWith/OpenWithButton.js [hunk 5]
```diff
@@ -29,32 +29,53 @@ type Props = {
...
+        Icon = ICON_FILE_MAP[displayName];
+    } else {
+        Icon = IconOpenWith;
+    }
 
     return (
-        <Tooltip text={getTooltip(displayDescription, error, isLoading)} position="bottom-center">
-            <Button isDisabled={isDisabled} onClick={() => onClick(displayIntegration)}>
+        <Tooltip
+            text={getTooltip(isLoading, error, isDisplayIntegrationDisabled, disabledReasons, displayDescription)}
```
Comment (jstoffan) [inline src/components/ContentOpenWith/OpenWithButton.js]: Consider alphabetizing or grouping params together, like `isLoading` with `isDisabled`.

File: src/components/ContentOpenWith/OpenWithButton.js [hunk 6]
```diff
@@ -29,32 +29,53 @@ type Props = {
...
+    if (isDisabled) {
+        return disabledReasons[0] || <FormattedMessage {...messages.emptyOpenWithDescription} />;
     }
+
     if (displayDescription) {
         return displayDescription;
     }
 
-    return <FormattedMessage {...messages.emptyOpenWithDescription} />;
+    return disabledReasons[0] || <FormattedMessage {...messages.emptyOpenWithDescription} />;
```
Comment (priyajeet) [inline src/components/ContentOpenWith/OpenWithButton.js]: I would reduce the number of returns here specially considering some are returning something similar. The lesser the number of exit points from your function, less break points you/someone will have to put to debug what the final value ends up being. You could just figure out what message to show and at the end 

if (error) message = messages.errorOpenWithDescription;
else if (isDisabled) message = messages.emptyOpenWithDescription;
else message = messages.emptyOpenWithDescription

`return disabledReasons[0] || <FormattedMessage {...message} />`

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 7]
```diff
@@ -194,7 +203,57 @@ class ContentOpenWith extends PureComponent<Props, State> {
...
+            this.boxEdit = new BoxEdit();
+            // Check if Box Tools is installed and reachable
+            this.boxEdit
+                .checkBoxEditAvailability()
+                .catch(() => {
+                    boxEditIntegration.isDisabled = true;
+                    boxEditIntegration.disabledReasons = ['Install Box Tools to open this file on your desktop'];
+                })
+                .then(this.getFileExtension)
+                .then(extension => {
```
Comment (priyajeet) [inline src/components/ContentOpenWith/ContentOpenWith.js]: Can this be migrated to another top level function, so that >>> arrowness reduces?

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 8]
```diff
@@ -194,7 +203,57 @@ class ContentOpenWith extends PureComponent<Props, State> {
...
+                    // Add the extension to render the correct icon.
+                    boxEditIntegration.extension = extension;
+
+                    if (boxEditIntegration.isDisabled) {
```
Comment (priyajeet) [inline src/components/ContentOpenWith/ContentOpenWith.js]: Should this check be made into a function, because wouldn't you want it in this.getFileExtension also? To prevent a possible XHR?

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 9]
```diff
@@ -298,6 +364,30 @@ class ContentOpenWith extends PureComponent<Props, State> {
...
+     * @param {ExecuteAPI} executeData - API response on how to open an executed integration
+
+     * @return {void}
+     */
+    executeBoxEditSuccessHandler = (executeData: ExecuteAPI): void => {
+        const { fileId, token, onExecute } = this.props;
+        const { url } = executeData;
+        const authCode = url.split(AUTH_CODE_DELIMITER)[1];
+
+        this.boxEdit.openFile(`f_${fileId}`, {
```
Comment (priyajeet) [inline src/components/ContentOpenWith/ContentOpenWith.js]: Since box edit is considered an API, shouldn't this be inside the api layer?

File: src/components/ContentOpenWith/IconFileMap.js [hunk 10]
```diff
@@ -3,13 +3,15 @@ import IconFileDefault from 'box-react-ui/lib/icons/file/IconFileDefault';
...
 import IconGoogleSlides from 'box-react-ui/lib/icons/google-docs/IconGoogleSlides';
+import FileIcon from 'box-react-ui/lib/icons/file-icon/FileIcon';
 
 const ICON_FILE_MAP = {
     default: IconFileDefault,
     'Adobe Sign': IconAdobeSign,
     'Google Docs': IconGoogleDocs,
     'Google Sheets': IconGoogleSheets,
     'Google Slides': IconGoogleSlides,
+    Open: FileIcon,
```
[the remark this answers is not in the document]
Comment (jeremypress) [reply src/components/ContentOpenWith/IconFileMap.js]: 😢 yeah default doesn't take an extension, this one does.

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 11]
```diff
@@ -194,7 +203,57 @@ class ContentOpenWith extends PureComponent<Props, State> {
      * @return {void}
      */
     fetchOpenWithSuccessHandler = (integrations: Array<Integration>): void => {
-        this.setState({ integrations, isLoading: false });
+        const boxEditIntegration = integrations.find(integration => {
+            return this.isBoxEditIntegration(integration);
+        });
+
+        if (boxEditIntegration && !boxEditIntegration.isDisabled) {
+            this.boxEdit = new BoxEdit();
```
[the remark this answers is not in the document]
Comment (jeremypress) [reply src/components/ContentOpenWith/ContentOpenWith.js]: Box Edit is already preserving an instance if one is previously created.

```   
let BoxEditInstance
 constructor() {
        if (!(BoxEditInstance instanceof BoxEdit)) {
            BoxEditInstance = this;
        }
        this.extensionRequestQueue = new Map();
        return BoxEditInstance;
    }```

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 12]
```diff
@@ -224,8 +283,17 @@ class ContentOpenWith extends PureComponent<Props, State> {
...
      * @return {void}
      */
-    onIntegrationClick = ({ appIntegrationId, displayName }: Integration): void => {
+    onIntegrationClick = (integration: Integration): void => {
+        const { appIntegrationId, displayName } = integration;
         const { fileId }: Props = this.props;
+
+        this.executeId = appIntegrationId;
+        if (this.isBoxEditIntegration(integration)) {
+            this.api
```
[the remark this answers is not in the document]
Comment (jeremypress) [reply src/components/ContentOpenWith/ContentOpenWith.js]: I'm using a different successhandler. I could try to combine but we do very different things if it's box edit vs. an online integration.

File: src/components/ContentOpenWith/ContentOpenWith.js (hunk 9 above)
Comment (jeremypress) [reply src/components/ContentOpenWith/ContentOpenWith.js]: do you mean adding it to the factory? I don't want to refactor the library itself yet since we don't own it.

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 13]
```diff
@@ -175,6 +180,10 @@ class ContentOpenWith extends PureComponent<Props, State> {
         }
     }
 
+    isBoxEditIntegration(integrationId: ?string): boolean {
```
Comment (jeremypress) [inline src/components/ContentOpenWith/ContentOpenWith.js]: add doc strings

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 14]
```diff
@@ -194,7 +209,73 @@ class ContentOpenWith extends PureComponent<Props, State> {
...
+    };
+
+    /**
+     * Uses Box Edit to check if Box Tools can open a given file type
+     *
+     * @param {Integration} boxEditIntegration - A Box Edit or Box Edit SFC integration
+     * @return {void}
+     */
+    canOpenExtensionWithBoxEdit = (integration: Integration): Promise<Integration> => {
+        const { extension } = integration;
```
[the remark this answers is not in the document]
Comment (jeremypress) [reply src/components/ContentOpenWith/ContentOpenWith.js]: I need the whole integration to do the pass through below.

File: src/components/ContentOpenWith/IconFileMap.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (jeremypress) [reply src/components/ContentOpenWith/IconFileMap.js]: done.

File: src/components/ContentOpenWith/OpenWithButton.js [hunk 15]
```diff
@@ -29,32 +29,41 @@ type Props = {
...
-        return <FormattedMessage {...messages.errorOpenWithDescription} />;
-    }
-    if (displayDescription) {
-        return displayDescription;
+
+    let message = <FormattedMessage {...messages.emptyOpenWithDescription} />;
+    if (disabledReasons.length > 0) {
+        [message] = disabledReasons;
+    } else if (error) {
+        message = <FormattedMessage {...messages.errorOpenWithDescription} />;
```
Comment (JustinHoldstock) [inline src/components/ContentOpenWith/OpenWithButton.js]: This seems like an unnecessary statement, considering it's identical, above

Comment (jeremypress) [reply src/components/ContentOpenWith/OpenWithButton.js]: What's it identical to? Here we're using the error message but we default to the empty message on L42.

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 16]
```diff
@@ -194,7 +209,82 @@ class ContentOpenWith extends PureComponent<Props, State> {
...
+                this.setState({ integrations, isLoading: false });
+            });
+    };
+
+    /**
+     * Fetches the file extension of the current file.
+     *
+     * @return {void}
+     */
+    setIntegrationFileExtension = (): Promise<BoxItem> => {
```
[the remark this answers is not in the document]
Comment (jeremypress) [reply src/components/ContentOpenWith/ContentOpenWith.js]: yep, it was previously modifying the object but now it's not!

File: src/api/File.js [hunk 17]
```diff
@@ -189,6 +190,24 @@ class File extends Item {
...
+
+    /**
+     * Gets the extension of a box file.
+     *
+     * @param {string} id - File id
+     * @param {Function} successCallback - Function to call with results
+     * @param {Function} errorCallback - Function to call with errors
+     * @return {Promise}
+     */
+    async getFileExtension(id: string, successCallback: Function, errorCallback: ElementsErrorCallback): Promise<void> {
```
[the remark this answers is not in the document]
Comment (jeremypress) [reply src/api/File.js]: It resolves the `return;` case not ideal, we have a mix of promise based/callback based logic in the API layer.

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 18]
```diff
@@ -194,7 +209,82 @@ class ContentOpenWith extends PureComponent<Props, State> {
...
+            return;
+        }
+
+        // If Box Edit is present and enabled, we need to set its ability to locally open the given file
+        this.setIntegrationFileExtension()
+            .then(data => {
+                boxEditIntegration.extension = data.extension;
+                return boxEditIntegration;
+            })
+            .then(this.checkBoxEditAvailability)
```
[the remark this answers is not in the document]
Comment (jeremypress) [reply src/components/ContentOpenWith/ContentOpenWith.js]: giving it a shot. Also finally isn't supported on IE so this could clean things up considerably.

File: src/api/APIFactory.js [hunk 19]
```diff
@@ -510,6 +516,21 @@ class APIFactory {
...
     }
+
+    /**
+     * API for Box Edit
+     *
+     * @param {boolean} shouldDestroy - true if the factory should destroy before returning the call
+     * @return {BoxEditAPI} BoxEditAPI instance
+     */
+    getBoxEditAPI(shouldDestroy: boolean): BoxEditAPI {
+        if (shouldDestroy) {
```
Comment (jstoffan) [inline src/api/APIFactory.js]: Seems a little odd to pass this in but not destroy the BoxEditAPI itself. Maybe just omit the param entirely?

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 20]
```diff
@@ -187,16 +200,90 @@ class ContentOpenWith extends PureComponent<Props, State> {
...
+        }
+
+        try {
+            const { extension } = await this.getIntegrationFileExtension();
+            boxEditIntegration.extension = extension;
+        } catch (error) {
+            boxEditIntegration.disabledReasons.push(
+                <FormattedMessage {...messages.executeIntegrationOpenWithErrorHeader} />,
+            );
+            boxEditIntegration.isDisabled = true;
```
Comment (jstoffan) [inline src/components/ContentOpenWith/ContentOpenWith.js]: Seems like these two lines are duplicated three times, which isn't ideal.

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 21]
```diff
@@ -187,16 +200,90 @@ class ContentOpenWith extends PureComponent<Props, State> {
...
+            this.setState({ integrations, isLoading: false });
+            return;
+        }
+
+        // If Box Edit is present and enabled, we need to set its ability to locally open the given file
+        if (!(await this.checkBoxEditAvailability())) {
```
Comment (jstoffan) [inline src/components/ContentOpenWith/ContentOpenWith.js]: I personally find this very difficult to read. If not await check?

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 22]
```diff
@@ -187,16 +200,90 @@ class ContentOpenWith extends PureComponent<Props, State> {
...
+        );
+
+        if (!boxEditIntegration || boxEditIntegration.isDisabled) {
+            this.setState({ integrations, isLoading: false });
+            return;
+        }
+
+        try {
+            const { extension } = await this.getIntegrationFileExtension();
+            boxEditIntegration.extension = extension;
```
Comment (jstoffan) [inline src/components/ContentOpenWith/ContentOpenWith.js]: It seems like you could create a new object to contain your overrides, then spread it onto the `boxEditIntegration` object at the end of the method, avoiding duplication of the `setState` call.

File: src/components/ContentOpenWith/ContentOpenWith.js (hunk 21 above)
Comment (DanDeMicco) [inline src/components/ContentOpenWith/ContentOpenWith.js]: Can this be done in parallel to `getIntegrationFileExtension`?

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 23]
```diff
@@ -187,16 +200,90 @@ class ContentOpenWith extends PureComponent<Props, State> {
...
+        });
+    };
+
+    /**
+     * Uses Box Edit to check if Box Tools is installed and reachable
+     *
+     * @param {Integration} boxEditIntegration - A Box Edit or Box Edit SFC integration
+     * @return {void}
+     */
+    checkBoxEditAvailability = (): Promise<boolean> => {
```
Comment (jstoffan) [inline src/components/ContentOpenWith/ContentOpenWith.js]: The other similar method begins with `can` below. Since the promise here always resolves with a boolean, might be good to stay consistent.

File: src/components/ContentOpenWith/ContentOpenWith.js [hunk 24]
```diff
@@ -187,16 +200,90 @@ class ContentOpenWith extends PureComponent<Props, State> {
             .getOpenWithIntegrations(fileId, language, this.fetchOpenWithSuccessHandler, this.fetchErrorHandler);
     }
 
+    setResolvedIntegrationState = (integrations: Array<Integration>) =>
```
Comment (ConradJChan) [inline src/components/ContentOpenWith/ContentOpenWith.js]: Is this used anywhere?

File: src/components/ContentOpenWith/OpenWithButton.js [hunk 25]
```diff
@@ -29,32 +29,41 @@ type Props = {
...
-    const IntegrationIcon = displayName && ICON_FILE_MAP[displayName];
-    const Icon = IntegrationIcon || IconOpenWith;
+
+    const Icon = displayName ? getIcon(displayName) : IconOpenWith;
 
     return (
-        <Tooltip text={getTooltip(displayDescription, error, isLoading)} position="bottom-center">
-            <Button isDisabled={isDisabled} onClick={() => onClick(displayIntegration)}>
+        <Tooltip text={getTooltip(displayDescription, isLoading, error, disabledReasons)} position="bottom-center">
+            <Button isDisabled={isDisabled} onClick={() => onClick(displayIntegration, isLoading)}>
```
Comment (jstoffan) [inline src/components/ContentOpenWith/OpenWithButton.js]: Does `isLoading` need to be passed back to the click handler?

File: src/components/ContentOpenWith/ContentOpenWith.js (hunk 23 above)
Comment (jeremypress) [reply src/components/ContentOpenWith/ContentOpenWith.js]: switching to a similar format. `isBoxEditAvailable`
